File: summer-school/MPI/jacobi/jacobi-serial.f90
program jacobi

  implicit none

  real(8), parameter :: top=1.0,bottom=10.0, left=1.0,right=1.0
  integer, parameter :: max_iter=1000000
  integer, parameter :: nprint=100
  integer :: nx,ny
  integer :: i,j,k,iter

  real(8) :: norm,bnorm
  real(8) :: tol=1e-4
  real(8), allocatable, dimension(:,:) :: grid,grid_new 
  character(len=32) :: arg


  if (command_argument_count() /= 2) then
     print *, &
          "You must provide two command line arguments, the global size in X and the global size in Y"
     stop
  end if

  call get_command_argument(1, arg)
  read(arg,*) nx
  call get_command_argument(2, arg)
  read(arg,*) ny

  print *,'grid size ',nx, ' x ',ny
  allocate(grid(0:ny+1,0:nx+1),grid_new(0:ny+1,0:nx+1) )


  ! boundary conditions
  grid(0,:)=left
  grid(ny+1,:)=right
  grid(:,0) = top
  grid(:,nx+1)=bottom

  grid_new=grid

  ! starting conditions
  grid(1:ny,1:nx)=0.0

  !! initial norm value

  norm=0.0
  do j=1,nx
     do i=1,ny
        norm=norm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1))**2)
     enddo
  enddo

  bnorm=sqrt(norm)
  !print *,'bnorm=',bnorm

  do iter=1, max_iter

     norm=0.0
     do j=1,nx
        do i=1,ny
           norm=norm+((grid(i,j)*4-grid(i-1,j)-grid(i+1,j)-grid(i,j-1)-grid(i,j+1))**2)
        enddo
     enddo

     norm=sqrt(norm)/bnorm

     if (norm .lt. tol) exit

     do j=1,nx
        do i=1,ny
           grid_new(i,j)=0.25 * (grid(i-1,j) + grid(i+1,j) + grid(i,j-1) + grid(i,j+1))
        end do
     end do

     grid=grid_new

     if (mod(iter,nprint)==0) then
        write(*,*) 'Iteration ',iter,' Relative norm ',norm
     endif

  enddo
  write(*,*) 'Terminated on ',iter, ' iterations, relative norm=',norm

  deallocate(grid,grid_new)

end program jacobi

File: summer-school/MPI/jacobi/jacobi-serial.c
#include <stdio.h>
#include <math.h>

// Grid boundary conditions
#define RIGHT 1.0
#define LEFT 1.0
#define TOP 1.0
#define BOTTOM 10.0

// Algorithm settings
#define TOLERANCE 0.0001
#define NPRINT 1000
#define MAX_ITER 1000000

int main(int argc, char*argv[]) {

  int k;
  double tmpnorm,bnorm,norm;

  if (argc !=3) {
    printf("usage: $argv[0] GRIDX GRIDY\n");
      return(1);
  }

  int nx=atoi(argv[1]);
  int ny=atoi(argv[2]);
  int ny2=ny+2;

  printf("grid size %d X %d \n",ny,ny);
  double *grid= (double*)malloc(sizeof(double)*(nx+2)*(ny+2));
  double *grid_new= (double*)malloc(sizeof(double)*(nx+2)*(ny+2));
  double *temp= (double*)malloc(sizeof(double)*(nx+2)*(ny+2));

  // Initialise Grid boundaries
  int i,j;
  for (i=0;i<ny+2;i++) {
    grid_new[i]=grid[i]=TOP;
    j=(ny+2)*(nx+1)+i;
    grid_new[j]=grid[j]=BOTTOM;
  }
  for (i=1;i<nx+1;i++) {
    j=(ny+2)*i;
    grid_new[j]=grid[j]=LEFT;
    grid_new[j+ny+1]=grid[j+ny+1]=RIGHT;
  }
   
  // Initialise rest of grid
  for (i=1;i<=nx;i++) 
    for (j=1;j<=ny;j++) {
      k=(ny+2)*i+j;
      grid_new[k]=grid[k]=0.0;
   }
/*  for (i=0;i<=nx+1;i++) {
    for (j=0;j<=ny+1;j++){
      printf("->%lf ",grid[j+i*(ny+2)]);
    }
    printf("\n");
  }
*/

  tmpnorm=0.0;
  for (i=1;i<=nx;i++) {
    for (j=1;j<=ny;j++) {
      k=(ny+2)*i+j;            
      tmpnorm=tmpnorm+pow(grid[k]*4-grid[k-1]-grid[k+1] - grid[k-(ny+2)] - grid[k+(ny+2)], 2); 

    }
  }
  bnorm=sqrt(tmpnorm);

  int iter;
  for (iter=0; iter<MAX_ITER; iter++) {

    tmpnorm=0.0;
    for (i=1;i<=nx;i++) {
     for (j=1;j<=ny;j++) {
      k=(ny+2)*i+j;
      tmpnorm=tmpnorm+pow(grid[k]*4-grid[k-1]-grid[k+1] - grid[k-(ny+2)] - grid[k+(ny+2)], 2); 
    }
  }
    
    norm=sqrt(tmpnorm)/bnorm;
    //printf("tmpnorm=%lf norm=%lf\n",tmpnorm,norm);
    if (norm < TOLERANCE) break;

    for (i=1;i<=nx;i++) {
      for (j=1;j<=ny;j++) {
        k=(ny+2)*i+j;    
	grid_new[k]=0.25 * (grid[k-1]+grid[k+1] + grid[k-(ny+2)] + grid[k+(ny+2)]);
      }
    }
    memcpy(temp, grid_new, sizeof(double) * (nx + 2) * (ny+2));
    memcpy(grid_new, grid, sizeof(double) * (nx + 2) * (ny+2));
    memcpy(grid, temp, sizeof(double) * (nx + 2) * (ny+2));

    if (iter % NPRINT ==0) printf("Iteration =%d ,Relative norm=%e\n",iter,norm);
  }

  printf("Terminated on %d iterations, Relative Norm=%e \n", iter,norm);
  
//  for (i=0;i<=nx+1;i++) {
//    for (j=0;j<=ny+1;j++){
//     printf("->%lf ",grid[j+i*(ny+2)]);
//    }
//    printf("\n");
//  }

  free(grid);
  free(temp);
  free(grid_new);
  return 0;
    

  }
